align comments based on prev/next line

millermedeiros opened the pull request:
this is still WIP but creating the pull request so you can see the new behavior. all the logic was implemented on [rocambole-indent](https://github.com/millermedeiros/rocambole-indent) and we are just reusing the `indent.alignComments()` method

still need to update the tests to match the new behavior but want to confirm if that is really the expected behavior... specially for cases like:

```js
try {
  foo();
} catch (e) {
// this won't be indented since it follows next line
);
```

fixes #209
fixes #270

jzaefferer commented on test/compare/default/obj_expression-in.js:
```diff
@@ -85,8 +85,9 @@ Ext.define('VMS.model.User', {
+      // align with previous line because of line break
 
-      // lorem ipsum
+      // align with "}"
```
If `align with "}"` applied, shouldn't this comment and the two below have less indent?

jzaefferer replied:
Nevermind, looking at input file...

jzaefferer commented:
I've looked through all the updated tests. Looks good to me.

mgutz commented:
That's an unusual style. I would have expected the comment above to be indented.

For example

```
function foo() {
// I would expect this comment to be indented
}
```

millermedeiros commented:
I agree with @mgutz that it's weird and the rule that he described should be easy to implement and makes way more sense.

millermedeiros commented:
uhm, there are a few edge cases like:

```js
function foo() {
// this comment should be indented

  return bar();
}
```

I considered that it would add the indent only if it was inside an empty block when in fact it should add if comment is first non-empty token after `{` or `[` or `(`.

will need to release yet another version of rocambole-indent :/

millermedeiros commented:
I updated the `rocamboleIndent.alignComments()` method to handle all the edge cases that I could think of and decided to add a new option `indent.AlignComments` to toggle the behavior:

```js
// indent.AlignComments = true (default)
if (foo) {
  bar();
// aligned with "else" (new behavior)
} else {
  baz();
}
```

and:

```js
// indent.AlignComments = false
if (foo) {
  bar();
  // aligned with "bar();" (old behavior)
} else {
  baz();
}
```

I also added a setting `indent.CommentInsideEmptyBlock` that user can use to toggle between:

```js
// indent.CommentInsideEmptyBlock = 1 (default)
function noop() {
  // this will be indented
}
```

and:

```js
// indent.CommentInsideEmptyBlock = 0
function noop() {
// this won't be indented
}
```

jzaefferer commented:
This looks good to me. Hopefully this is the last time we need to update line comment handling!

millermedeiros commented:
squased commits and landed on master \o/
